pkg/controller/watermarkpodautoscaler/replica_calculator.go

```diff
@@ -190,36 +204,67 @@ func getReplicaCount(logger logr.Logger, currentReplicas int32, wpa *v1alpha1.Wa
 	return replicaCount, utilizationQuantity.MilliValue()
 }
 
-func groupPods(logger logr.Logger, podList *v1.PodList, metrics metricsclient.PodMetricsInfo, resource v1.ResourceName, delayOfInitialReadinessStatus time.Duration) (readyPods, ignoredPods sets.String) {
+func (c *ReplicaCalculator) getReadyPodsCount(namespace string, selector labels.Selector, readinessDelay time.Duration) (int32, error) {
+	podList, err := c.podLister.Pods(namespace).List(selector)
+	if err != nil {
+		return 0, fmt.Errorf("unable to get pods while calculating replica count: %v", err)
+	}
+
+	if len(podList) == 0 {
+		return 0, fmt.Errorf("no pods returned by selector while calculating replica count")
+	}
+
+	toleratedAsReadyPodCount := 0
+
+	for _, pod := range podList {
+		_, condition := getPodCondition(&pod.Status, corev1.PodReady)
+		if condition == nil || pod.Status.StartTime == nil {
+			log.V(4).Info("Pod unready", "namespace", pod.Namespace, "name", pod.Name)
+			continue
+		}
+		if pod.Status.Phase == corev1.PodRunning && condition.Status == corev1.ConditionTrue ||
+			// We only care about the time after start as a warm up. If a pod becomes unresponsive it will not elect for a readinessDelay tolerance.
+			// We do not use v1.ConditionFalse, because we only tolerate the Pending state stuck in the same condition for more than readinessDelay.
+			// Pending includes the time spent pulling images onto the host.
+			pod.Status.Phase == corev1.PodPending && condition.LastTransitionTime.Sub(pod.Status.StartTime.Time) < readinessDelay {
+			toleratedAsReadyPodCount++
+		}
+	}
+	return int32(toleratedAsReadyPodCount), nil
+}
+
+func groupPods(logger logr.Logger, podList []*corev1.Pod, metrics metricsclient.PodMetricsInfo, resource corev1.ResourceName, delayOfInitialReadinessStatus time.Duration) (readyPods, ignoredPods sets.String) {
 	readyPods = sets.NewString()
 	ignoredPods = sets.NewString()
-	for _, pod := range podList.Items {
+	missing := sets.NewString()
+	for _, pod := range podList {
 		// Failed pods shouldn't produce metrics, but add to ignoredPods to be safe
-		if pod.Status.Phase == v1.PodFailed {
+		if pod.Status.Phase == corev1.PodFailed {
```

Comment:
could you check that evicted pod is also considered as `PodFailed`?

Reply:
Thanks for pointing this out.
Yes, evicted pods have a Status.Phase == Failed

```
kubectl get pods --all-namespaces | grep Evicted
kube-system                metrics-server-XXX-6gv6s     0/2     Evicted     0    213d

kubectl get po metrics-server-XXX-6gv6s -n kube-system -oyaml
[...]
status:
  message: 'The node was low on resource: imagefs.'
  phase: Failed
  reason: Evicted
  startTime: "2019-11-05T02:51:25Z"
```